### PR TITLE
Gtk3 update6

### PR DESCRIPTION
**Pull request by Bob-IT**
This changes the register to use CSS for the background colours. the foreground colour will be what ever is currently used and the border colours are still hard coded, not sure if it worth changing this but your thoughts are appreciated.
Bob

Here is some CSS for your .gtk-3.0-gnucash.css file to illustrate what can be done...

.sheet {
  font: 15px arial, sans-serif;
}

.header-color {
  background-color: seagreen;
  color: white;
}

.primary-color {
  background-color: pink;
}

.primary-color:insensitive {
  background-color: plum;
}

.secondary-color {
  background-color: lightgreen;
}

.secondary-color:insensitive {
  background-color: plum;
}

.split-color {
  background-color: lightblue;
}

.split-color:insensitive {
  background-color: plum;
}

.cursor-color {
  background-color: #00BFFF;
  color: white;
}

### CODE REVIEW

**Comment by Bob-IT** [comment]
Just noticed I have failed to free one of the colours, will wait to see if further changes required before pushing change.
Bob

**Comment by Bob-IT** [comment]
There is some thing not quite right, please wait for an update before looking at this...
Bob

**Comment by Bob-IT** [comment]
Hopefully this is OK, all should work along with the CSS code above with the following possible changes...

cursor entry {
  background-color: #0099cc;
}
sheet {
  font: 20px arial, sans-serif;
}

The last commit to fix the missing hatchings required a change from using the gtkrc routines to the normal ones with changes. There is a possibility to combine getting the background CSS class and foreground but not sure if that would be good thing.

Will wait for feedback,
Bob

**Comment by kubrickfr** [comment]
This PR enables scrolling in the ledger, but whether I scroll up or down, it always scrolls up

**Comment by Bob-IT** [comment]
I think I was too restrictive on the scroll delta and it should of been less than or greater than 0 for direction, changed.

**Comment by kubrickfr** [comment]
Mmm. Still doesn't work for me **with a trackpad**, works fine with mouse wheel.

**Comment by kubrickfr** [comment]
Also, if the ledger is first clicked, the position of the scroll-bar is not updated when scrolling.

**Comment by Bob-IT** [comment]
This should fix the scrolling but not the trackpad as I do not have access to one for test. The scrolling looks like it may be down to this bug [(https://bugzilla.gnome.org/show_bug.cgi?id=765410)] but hiding and showing the scrollbar seems to get round it.
Bob

**Comment by kubrickfr** [comment]
@Bob-IT I'm sorry but it still doesn't work:
1) Trackpad scrolling up and down only works as intended when the scroll-bar has the focus, when the ledger has focus, scrolling up or down scrolls up (with the trackpad)
2) scroll-bar position is only refreshed when it has the focus.
(Using up to date arch linux)

**Comment by Bob-IT** [comment]
What version of GTK3 are you using ?

**Comment by kubrickfr** [comment]
extra/gtk3 3.22.21-1
I'm trying to add some debug output to help...

**Comment by Bob-IT** [comment]
Are you saying it does not work with a mouse or trackpad ?
I can not test a trackpad as I do not have access to one.
Try putting a couple of g_printf statements in gnucash-sheet.c gnucash_scroll_event to see if it is triggered and what the delta_y values are if you are looking at trackpad sheet scrolling.

**Comment by kubrickfr** [comment]
Not working with a trackpad and yes, that's exactly what I'm doing :)

**Comment by Bob-IT** [comment]
You could try adding this to line 2547 of gnucash-sheet.c GDK_SMOOTH_SCROLL_MASK to see if it makes any difference. This maybe of use (https://developer.gnome.org/gdk3/stable/gdk3-Events.html)

**Comment by kubrickfr** [comment]
Apoligies, I must have been working with a stale version of your patch or something, but when I rebuilt from scratch to add my debug it started to work.
So yes, all fix, all working, I can see the hiding/showing scroll bar in action.

**Comment by Bob-IT** [comment]
Glad it is working, just need to wait for the expert review which will no doubt raise some issues I did not think off.
Bob

File: gnucash/register/register-core/table-allgui.h [hunk 1]
```diff
@@ -182,89 +182,32 @@ struct table
...
+    COLOR_PRIMARY_FG,          // 10
+    COLOR_PRIMARY_FG_ACTIVE,   // 11
+    COLOR_SECONDARY_FG,        // 12
+    COLOR_SECONDARY_FG_ACTIVE, // 13
+    COLOR_SPLIT_FG,            // 14
+    COLOR_SPLIT_FG_ACTIVE,     // 15
 
     /* Other colors */
-    COLOR_NEGATIVE, /* Color to use for negative numbers */
+    COLOR_NEGATIVE,            // 16 Color to use for negative numbers
```
Comment (gjanssens) [inline gnucash/register/register-core/table-allgui.h]: A first preliminary scan of the uses of the enums suggests the differentiation between FG and BG colors is not used at all in the end. (It is still used to do the calculations, but [the only code that makes style changes based on these results](https://github.com/Bob-IT/gnucash/blob/f1d49047ab754bd938219754e5e9fe07438d0272/gnucash/register/register-gnome/gnucash-sheet.c#L2413) always checks for both FG and BG it one case statement)

So how about redefining the enum as follows:

```
typedef enum {
COLOR_UNDEFINED = 0,
COLOR_HEADER,
COLOR_PRIMARY,
COLOR_PRIMARY_ACTIVE,
COLOR_SECONDARY,
COLOR_SECONDARY_ACTIVE,
COLOR_SPLIT,
COLOR_SPLIT_ACTIVE,
COLOR_NEGATIVE,
} RegisterColor;
```

We can the probably do away with the enum arithmetic and reduce the code even more. But please double-check if my analysis is correct :)

File: gnucash/register/register-gnome/gnucash-register.c [hunk 2]
```diff
@@ -271,6 +271,11 @@ gnucash_register_sheet_resize (GnucashRegister *reg)
     // not get filled on load, this makes sure it does
     if (!reg->hscrollbar_visible)
         gtk_widget_queue_resize (GTK_WIDGET (reg->sheet));
+
+    // Make sure the sheet is the focus
+    if (!gtk_widget_has_focus(GTK_WIDGET (reg->sheet)))
+        gtk_widget_grab_focus (GTK_WIDGET (reg->sheet));
+
```
Comment (gjanssens) [inline gnucash/register/register-gnome/gnucash-register.c]: ... which eventually should become the item-edit if we manage to get the code cleaned up sufficiently.

**Comment by gjanssens** [comment]
I'm revisiting this after a good night sleep. If you haven't done so already, don't bother squashing the whitespace commits together. I believe your commits are clean, clear and well-contained, so squashing those still would just be a nitpick. No need to spend extra effort on that in this bigger work. If you did already, sorry I didn't realize this earlier and don't bother reverting though. I don't want to have you spend any more precious time on it!

The other comments are still valid though.

**Comment by Bob-IT** [comment]
OK, I was having trouble with squashing them, it kept telling me there was another git process running.
I have just rebased and force pushed as a start point. I will have some time tomorrow to make some changes so if you want to wait till then to grab it.
Bob

File: gnucash/register/register-gnome/gnucash-item-edit.c [hunk 3]
```diff
@@ -334,6 +334,57 @@ unblock_toggle_signals(GncItemEdit *item_edit)
...
+    
+    // Now draw a vertical line
+    cairo_set_source_rgb (cr, fg_color->red, fg_color->green, fg_color->blue);
+    cairo_set_line_width (cr, 1.0);
+    cairo_move_to (cr, cursor_x + 0.5, 2);
+    cairo_rel_line_to (cr, 0, height - 4);
+    cairo_stroke (cr);
+
+    return FALSE;
+}
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/register-gnome/gnucash-item-edit.c]: I started with the item_edit as the focus but there were so many other bits that wanted to start from the sheet that it made sense to get it working that way and then maybe think about changing it later.

File: gnucash/gnome-utils/ui/gnucash.css [hunk 4]
```diff
@@ -10,3 +10,87 @@ cursor entry {
...
 }
+
+/* Register defaults */
+@define-color register_header_bg_color #96B183;
+@define-color register_primary_bg_color #BFDEB9;
+@define-color register_secondary_bg_color #F6FFDA;
+@define-color register_split_bg_color #EDE7D3;
+@define-color register_cursor_bg_color #FFEF98;
+
+.app-header {
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/gnome-utils/ui/gnucash.css]: The first thing I thought of, I will change to register-...

File: gnucash/gnome-utils/ui/gnucash.css [hunk 5]
```diff
@@ -10,3 +10,87 @@ cursor entry {
...
+  background-color: @register_header_bg_color;
+}
+
+.app-header-dark {
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/gnome-utils/ui/gnucash.css]: No worries, gave it a go, possibly could tweak it but as you say that is what the preference is for, will remove.

File: gnucash/register/register-gnome/gnucash-sheet.c [hunk 6]
```diff
@@ -1055,10 +1055,17 @@ gnucash_sheet_draw_cb (GtkWidget *widget, cairo_t *cr, G_GNUC_UNUSED gpointer da
...
     gboolean result; //FIXME
 
     gtk_widget_get_allocation(widget, &alloc);
-    gtk_render_background (context, cr, alloc.x, alloc.y, alloc.width, alloc.height);
+
+    gtk_style_context_add_class (context, GTK_STYLE_CLASS_BACKGROUND);
+    gtk_render_background (context, cr, 0, 0, alloc.width, alloc.height);
+    gtk_style_context_get_color (context, GTK_STATE_FLAG_NORMAL, &color);
+    sheet->dark_theme = gnc_is_dark_theme (&color);
+    gtk_style_context_remove_class (context, GTK_STYLE_CLASS_BACKGROUND);
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/register-gnome/gnucash-sheet.c]: Not sure, I have in other places will need to check why not here.

File: gnucash/register/register-gnome/gnucash-sheet.c [hunk 7]
```diff
@@ -2321,104 +2328,69 @@ gnucash_sheet_realize_entry (GnucashSheet *sheet, GtkWidget *entry)
...
     {
     default:
-        return gdk_rgba_copy (&gn_white);
-
     case COLOR_UNKNOWN_BG:
-        return gdk_rgba_copy (&gn_white);
-
     case COLOR_UNKNOWN_FG:
-        return gdk_rgba_copy (&gn_black);
+        gtk_style_context_add_class (stylectxt, GTK_STYLE_CLASS_BACKGROUND);
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/register-gnome/gnucash-sheet.c]: Not sure, will check but I do not think it matters, not sure it ever gets used.

File: gnucash/register/register-gnome/gnucash-sheet.c [hunk 8]
```diff
@@ -262,6 +262,68 @@ gnucash_sheet_deactivate_cursor_cell (GnucashSheet *sheet)
...
+    }
+
+    result = pango_layout_xy_to_index (layout,
+                 PANGO_SCALE * (sheet->button_x - rect.x - x_offset),
+                 PANGO_SCALE * (h/2), &index, &trailing);
+
+    g_object_unref (layout);
+
+    return index + trailing;
+}
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/register-gnome/gnucash-sheet.c]: Yes, if the GtkEntry was used properly then the cursor would be there all ready.

File: gnucash/register/ledger-core/split-register-layout.c [hunk 9]
```diff
@@ -737,8 +737,8 @@ gnc_split_register_layout_add_cells (SplitRegister *reg,
                            QUICKFILL_CELL_TYPE_NAME,
                            N_("sample:Memo field sample text string") + 7,
                            CELL_ALIGN_LEFT,
-                           FALSE,
-                           TRUE);
+                           TRUE,
+                           FALSE);
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/ledger-core/split-register-layout.c]: Drag the right hand separator of the balance column to the right and the column will be shown on some registers, I think the basic account is one and credit card one.

File: gnucash/register/register-gnome/gnucash-sheet-private.c [hunk 10]
```diff
@@ -551,7 +551,7 @@ draw_cell (GnucashSheet *sheet,
         break;
 
     case CELL_ALIGN_RIGHT:
-        x_offset = width - 2 * CELL_HPADDING - logical_rect.width;
+        x_offset = width - 2 * CELL_HPADDING - logical_rect.width - 3;
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/register/register-gnome/gnucash-sheet-private.c]: Not sure where the difference comes from, maybe to do with borders, will see if I can find out.

File: gnucash/register/register-core/table-allgui.h (hunk 1 above)
Comment (Bob-IT) [reply gnucash/register/register-core/table-allgui.h]: So I think this answers my question about keeping foreground and background cell functions

File: gnucash/register/register-gnome/gnucash-sheet-private.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (gjanssens) [reply gnucash/register/register-gnome/gnucash-sheet-private.c]: For clarity, this code is not wrong. It was more a general observation that we still have much hardcoded sizing going on in the register code. But don't spend too much time on trying to eliminate that yet. We risk ending up rewriting our own implementation of a GtkTreeView if we push it too far. And that's definitely not something we want to do this late in the release cycle!

File: gnucash/register/register-gnome/gnucash-sheet-private.c [hunk 11]
```diff
@@ -707,8 +707,9 @@ gnucash_sheet_draw_cursor (GnucashCursor *cursor, cairo_t *cr)
     cairo_stroke (cr);
 
     // make the bottom line thicker
-    cairo_move_to (cr, cursor->x - x + 0.5, cursor->y - y + cursor->h - 1.5);
+    cairo_move_to (cr, cursor->x - x + 0.5, cursor->y - y + cursor->h - 3.5);
     cairo_rel_line_to (cr, cursor->w, 0);
+    cairo_set_line_width (cr, 1.0);
```
[the remark this answers is not in the document]
Comment (gjanssens) [reply gnucash/register/register-gnome/gnucash-sheet-private.c]: I understood the reason you decided to change it. And conceptually it's right (add a highlight marker inside the cell that's highlighted). However there are several drawbacks, all of them due to the register's choice of tight packing (we have almost no padding between rows).
1. as said the line is missing in the cell that's highlighted. This is most disturbing in an empty cell, such as a Num cell. But it's equally distracting in other cells.
2. As we have so little padding adding the line inside the row causes some characters to partially disappear in that line. This is most notable with characters like 'g', 'q', 'y' and to some extend also with ','.

So if we want to add the line inside the highlighted row, we will have to rethink our padding. That is in that case we'll need to make our cells 1px taller (for the full table, not only the hightlighted row).

Alternatively we add the highlight line below the highlighted row (so in fact at the top of the next cell) and compensate for this when scrolling by adjusting the scroll with one additional pixel. I don't know if that's easily doable tbh.

**Comment by gjanssens** [comment]
On donderdag 28 september 2017 16:27:22 CEST Bob-IT wrote:
> Bob-IT commented on this pull request.
> 
> 
> 
> 
>      /* Other colors */
> -    COLOR_NEGATIVE, /* Color to use for negative numbers */
> +    COLOR_NEGATIVE,            // 16 Color to use for negative numbers
> 
> So I think this answers my question about keeping foreground and background
> cell functions

Remind me when/where you asked about that?

**Comment by Bob-IT** [comment]
Forth comment from the top worded slightly differently.

File: gnucash/register/register-core/table-allgui.h (hunk 1 above)
Comment (gjanssens) [reply gnucash/register/register-core/table-allgui.h]: So indeed it does answer your question :) Good you had already spotted it.

**Comment by Bob-IT** [comment]
Geert,
I think these are the main changes requested, mostly visible so if OK grab them and close this PR
Will look at the other changes in a new branch.
Still think about the text alignment, on my screen a non focused cell with an uppercase T, it looks like 6 pixels above and 4 pixels below to the border so there might be room to move the text up.

The last commit removes the use of CSS files to a Gresource file and I have introduced another CSS file for Gtk3.20 or greater that fixes the pseudo-class due to deprecated class. The changes to the  Cmake files and Autotools files is the best I could come up with so if there is a better way please change.

Bob

**Comment by gjanssens** [comment]
On vrijdag 29 september 2017 15:33:53 CEST Bob-IT wrote:
> Geert,
> I think these are the main changes requested, mostly visible so if OK grab
> them and close this PR Will look at the other changes in a new branch.
> Still think about the text alignment, on my screen a non focused cell with
> an uppercase T, it looks like 6 pixels above and 4 pixels below to the
> border so there might be room to move the text up.
> 
That in itself is an interesting observation, because on my system there's 
exactly 3 px above and 3px below an uppercase T. Can you post a screenshot 
somewhere ?


Other than that I have pushed your changes, so thanks a lot!

**Comment by jralls** [comment]
Is this PR complete?

**Comment by Bob-IT** [comment]
I am closing this as any follow up changes will be done in new PR's.

**Pull request closed without merging**